activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -209,10 +209,12 @@ def in_order_of(key, series, filter: true)
   #   Set.new.sole        # => Enumerable::SoleItemExpectedError: no item found
   #   { a: 1, b: 2 }.sole # => Enumerable::SoleItemExpectedError: multiple items found
   def sole
-    case count
-    when 1   then return first # rubocop:disable Style/RedundantReturn
+    first_two = first(2)
+
+    case first_two.count
+    when 1   then first_two.first
     when 0   then raise ActiveSupport::EnumerableCoreExt::SoleItemExpectedError, "no item found"
-    when 2.. then raise ActiveSupport::EnumerableCoreExt::SoleItemExpectedError, "multiple items found"
+    when 2   then raise ActiveSupport::EnumerableCoreExt::SoleItemExpectedError, "multiple items found"
```

Comment:
This causes array allocation on `first(2)`. Maybe a better implementation memory wise would be:

``` ruby
result = none = Object.new
each do |element|
  if result == none
    result = element
  else
    raise ActiveSupport::EnumerableCoreExt::SoleItemExpectedError, "multiple items found"
end
if result == none
  ActiveSupport::EnumerableCoreExt::SoleItemExpectedError, "no item found"
else
  result
end
```

Reply:
This causes object allocation (`Object.new`), which is the same. Can be extracted into a constant of course, but I would prefer simplicity over complexity in not performance critical code.

Reply:
The solution proposed at the end of https://github.com/rails/rails/issues/54335#issue-2807665461 doesn't allocate and is relatively simple too, but up to you.